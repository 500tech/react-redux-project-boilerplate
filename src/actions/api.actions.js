// @flow
import { set } from 'lodash/fp';
import type { ApiAction } from 'types/redux.types';

export const apiAction = (action: ApiAction): ApiAction =>
  set('meta.api', true, action);
