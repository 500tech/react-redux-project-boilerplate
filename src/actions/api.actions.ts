import { BaseAction } from 'types/base-redux.types';

export type HttpMethod =
  | 'get'
  | 'post'
  | 'put'
  | 'delete'
  | 'path'
  | 'head'
  | 'options';

type OnSuccess<ResponseBody> = (data: ResponseBody) => BaseAction;
type OnError = (error: any) => void;

export interface ApiAction<ResponseBody> extends BaseAction {
  meta: {
    api: true;
  };
  payload: {
    networkLabel: string;
    method: HttpMethod;
    path: string;
    data?: any;
    baseUrl?: string;
    onSuccess: OnSuccess<ResponseBody> | OnSuccess<ResponseBody>[];
    onError?: OnError | OnError[];
  };
}
