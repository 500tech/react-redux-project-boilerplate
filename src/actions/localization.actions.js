// @flow
// TODO: remove this file if no localization is needed
import type { BaseAction } from 'types/redux.types';

export const SET_LOCALE: string = '[localization] Set Locale';

export type SetLocaleAction = BaseAction<{|
  +locale: string
|}>;
export type LocalizationAction = SetLocaleAction;

export const setLocale = (locale: string): SetLocaleAction => ({
  type: SET_LOCALE,
  payload: {
    locale
  }
});
