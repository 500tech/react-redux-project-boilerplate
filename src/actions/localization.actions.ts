// TODO: remove this file if no localization is needed
import { BaseAction } from 'types/base-redux.types';

export const SET_LOCALE: string = '[localization] Set Locale';

export interface SetLocaleActionPayload {
  locale: string;
}

export type SetLocaleAction = BaseAction<SetLocaleActionPayload>;
export type LocalizationAction = SetLocaleAction;

export const setLocale = (locale: string): SetLocaleAction => ({
  type: SET_LOCALE,
  payload: {
    locale
  }
});
