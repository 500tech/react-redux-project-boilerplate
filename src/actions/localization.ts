// TODO: remove this file if no localization is needed
import { BaseAction } from 'types/base-redux';
import { LocaleTypes } from 'constants/locales';

export const SET_LOCALE: string = '[localization] Set Locale';

export interface SetLocaleAction extends BaseAction {
  payload: {
    locale: string;
  };
}

export const setLocale = (locale: LocaleTypes) => ({
  type: SET_LOCALE,
  payload: {
    locale
  }
});
