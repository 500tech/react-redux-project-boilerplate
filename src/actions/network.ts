import { BaseAction } from 'types/base-redux';

export const START_NETWORK: string = '[network] Start';
export const END_NETWORK: string = '[network] End';

export interface NetworkAction extends BaseAction {
  payload: {
    networkLabel: string;
  };
}

export const startNetwork = (networkLabel: string) => ({
  type: START_NETWORK,
  payload: {
    networkLabel
  }
});

export const endNetwork = (networkLabel: string): NetworkAction => ({
  type: END_NETWORK,
  payload: {
    networkLabel
  }
});
