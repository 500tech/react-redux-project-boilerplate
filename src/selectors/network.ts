import { createSelector } from 'reselect';
import { some, castArray } from 'lodash/fp';

import { State } from 'types/redux';

// TODO: make into selector factory
export const makeIsLoadingSelector = (networkLabel: string | string[]) =>
  createSelector(
    (state: State) => state.network,
    (network) => {
      const labels = castArray(networkLabel);

      return some((currentLabel) => network[currentLabel] > 0, labels);
    }
  );
