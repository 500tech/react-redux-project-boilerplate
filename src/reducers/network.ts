import { set } from 'lodash/fp';
import { handleActions } from 'redux-actions';

import * as AT from 'actions/network';

export type NetworkState = Record<string, number>;

const initialState: NetworkState = {};

const network = handleActions<NetworkState>(
  {
    [AT.START_NETWORK](state, action: ReturnType<typeof AT.startNetwork>) {
      const { networkLabel } = action.payload;

      return set(
        networkLabel,
        state[networkLabel] ? state[networkLabel] + 1 : 1,
        state
      );
    },
    [AT.END_NETWORK](state, action: ReturnType<typeof AT.endNetwork>) {
      const { networkLabel } = action.payload;

      return set(
        networkLabel,
        state[networkLabel] ? state[networkLabel] - 1 : 0,
        state
      );
    }
  },
  initialState
);

export default network;
