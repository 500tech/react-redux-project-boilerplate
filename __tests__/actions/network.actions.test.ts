import { startNetwork, endNetwork } from 'actions/network.actions';

describe('Actions: network', () => {
  test('startNetwork', () => {
    expect(startNetwork('posts')).toMatchSnapshot();
  });

  test('endNetwork', () => {
    expect(endNetwork('posts')).toMatchSnapshot();
  });
});
