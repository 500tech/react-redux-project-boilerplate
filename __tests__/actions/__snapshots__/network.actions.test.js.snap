// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Actions: network endNetwork 1`] = `
Object {
  "payload": Object {
    "label": "posts",
  },
  "type": "[network] End",
}
`;

exports[`Actions: network startNetwork 1`] = `
Object {
  "payload": Object {
    "label": "posts",
  },
  "type": "[network] Start",
}
`;
